bar_app/serializers.py

```diff
@@ -0,0 +1,169 @@
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+from django.contrib.auth.models import User
+from .models import *
+
+class UserSerializer(serializers.ModelSerializer):
+    """
+    Serializer for User model.
+    """
+    class Meta:
+        model = User
+        fields = ["username", "password"]
+        extra_kwargs = {"password": {"write_only": True}}
+
+class ReferenceSerializer(serializers.ModelSerializer):
+    """
+    Serializer for Reference model.
+    """
+    availability = serializers.SerializerMethodField()
+
+    class Meta:
+        model = Reference
+        fields = "__all__"
+        read_only_fields = ["id"]
+
+    def validate_ref(self, value):
+        """
+        Check if the ref exists in the database.
+        """
+        if Reference.objects.filter(ref=value).exists():
+            raise serializers.ValidationError("This reference already exists.")
+        return value
+
+    def validate_name(self, value):
+        """
+        Validate the data for creating a new reference.
+        """
+        if not value.strip():
+            raise serializers.ValidationError("Name field cannot be empty.")
+
+        return value
+
+    def get_availability(self, instance):
+        stocks = Stock.objects.filter(reference=instance)
+
+        if not stocks.exists():
+            return "outofstock"
+
+        for stock in stocks:
+            if stock.stock > 0:
+                return "available"
+        return "outofstock"
```

Comment:
Although this answer initial demand, this isn't the most optimized approach.
This function will be executed at each instance serialization, meaning you are hitting the database everytime the serializer gets called.
Could it be done by annotating queryset on view instead ?